<div align="center" style="border: 2px solid #E57373; border-radius: 10px; padding: 20px; background-color: #FFCDD2;">
    <h1 style="font-size: 28px; color: #D32F2F;">Clustering mediante Algoritmos de Embedding de Grafos para el Análisis de Postulantes en el Proceso de Admisión de UNIQ</h1>
    <p style="font-size: 18px; color: #EF6C00;">Explorando Perfiles y Tendencias a través de Datos Transformados</p>
    <img src="../images/logo.png" alt="Imagen de ejemplo" style="border-radius: 5px; box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2); max-width: 80%; margin-top: 20px;">
</div>


# <a id="1"><font size="6"><div style="border-radius:5px;color:white;background-color:#FA8072;">  1. Introducción 🚀</div> </font></a>


<div style="background-color: #dbcfce; padding: 10px; border-radius: 5px; box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.1);">
    <p style="font-size: 16px; line-height: 1.5;color: #000;">
        En la última década, los gráficos de conocimiento (KG) han surgido como una área destacada en la IA. Un KG es esencialmente un gráfico multirelacional, etiquetado y dirigido, con semántica. El aumento en la publicación de datos estructurados en la web y el éxito de proyectos como Google Knowledge Graph y Amazon Product Graph han impulsado esta tendencia. Además, la naturaleza interdisciplinaria de la investigación académica ha sido clave. En este contexto, esta revisión panorámica de la investigación de KG busca sintetizar las diferentes líneas de investigación, resaltando cómo se unen en comunidades como el procesamiento del lenguaje natural, bases de datos y la web semántica.
    </p>
</div>


## <font size="6"><div style="border-radius:10px;border: 2px solid #DEB887;color:white;padding: 20px;background-color:#22938d;"> 🎯 Objetivo del Proyecto</div> </font>


<div style="background-color: #dbcfce; padding: 10px; border-radius: 5px; box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.1);">
    <p style="font-size: 16px; line-height: 1.5;color: #000;">
        El objetivo principal de este proyecto es aplicar técnicas de análisis de datos avanzadas para realizar un estudio exhaustivo de los postulantes en el proceso de admisión de la UNIQ. Se busca implementar algoritmos de clustering para agrupar a los postulantes en categorías que permitan identificar perfiles y tendencias comunes. Además, se empleará el concepto de embedding para transformar los datos en una representación más compacta y significativa, lo que facilitará la detección de similitudes y diferencias entre los estudiantes.
    </p>
</div>


## <font size="6"><div style="border-radius:10px;border: 2px solid #DEB887;color:white;padding: 20px;background-color:#22938d;"> 🎯 Contexto y Motivacion</div> </font>


<div style="border: 2px solid #3498DB; border-radius: 10px; padding: 20px; background-color: #E6F7FF;color: black">
    <h2 style="color: #2980B9;">🔍 Enfoque del Proyecto</h2>
    <p>
        El presente proyecto de investigación tiene su atención centrada en el análisis de la "Lista de Postulantes del Proceso de Admisión 2019 al 2023 de la Universidad Nacional Intercultural de Quillabamba" (UNIQ). El objetivo fundamental es desvelar patrones y relaciones ocultas entre los datos de los postulantes a través de técnicas especializadas como clustering, embedding y el algoritmo de la araña.
    </p>
    <hr style="border: 1px solid #3498DB; margin: 15px 0;">
    <p>
        Se pretende implementar algoritmos de clustering para agrupar a los postulantes en categorías que permitan identificar perfiles y tendencias comunes. Adicionalmente, se empleará el concepto de embedding para transformar los datos en una representación más compacta y significativa, facilitando así la identificación de similitudes y diferencias entre los estudiantes.
    </p>
    <hr style="border: 1px solid #3498DB; margin: 15px 0;">
    <p>
        Dada la diversidad de variables presentes en los datos, que abarcan desde información geográfica y demográfica hasta aspectos académicos y lingüísticos, es crucial optimizar los procesos de selección y agrupación de los estudiantes. Esta optimización podría permitir la identificación de segmentos de estudiantes con características similares. Estos segmentos podrían ofrecer información valiosa para mejorar políticas de admisión, personalizar la experiencia educativa y diseñar estrategias de apoyo estudiantil.
    </p>
</div>


## <font size="6"><div style="border-radius:10px;border: 2px solid #DEB887;color:white;padding: 20px;background-color:#22938d;"> 🎯 Descripción general de los datos y la problemática

</div> </font>


<div style="border: 2px solid #3498DB; border-radius: 10px; padding: 20px; background-color: #E6F7FF;color: black;">
    <h2 style="color: #2980B9;">📊 Conjunto de Datos: Lista de Postulantes del Proceso de Admisión 2019 al 2023</h2>
    <p>
        El conjunto de datos <em>"Lista de Postulantes del Proceso de Admisión 2019 al 2023 de la Universidad Nacional Intercultural de Quillabamba"</em> es proporcionado por la Universidad Nacional Intercultural de Quillabamba. Contiene información detallada sobre los individuos que participaron en el proceso de admisión de la universidad durante los años 2019 al 2023.
    </p>
    <p>
        Este conjunto de datos cubre un período de cinco años y ofrece una amplia variedad de atributos que describen diferentes aspectos de los postulantes.
    </p>
    <hr style="border: 1px solid #3498DB; margin: 15px 0;">
    <p>
        La información se presenta en formato tabular, donde cada fila representa un postulante individual, y las columnas contienen datos relacionados con varios aspectos de la postulación y el 
        perfil del estudiante. Los atributos incluyen categorías como:
    </p>
    <ul>
        <li>Información geográfica (ubicación de nacimiento y residencia)</li>
        <li>Detalles académicos (escuela de postulación, modalidad, año y lugar de culminación de estudios)</li>
        <li>Características personales (edad, sexo)</li>
        <li>Habilidades lingüísticas (segunda lengua, lectura, escritura, habla)</li>
        <li>Información financiera (pago de admisión)</li>
        <li>Detalles sobre discapacidad</li>
        <li>Pertenencia a comunidades específicas</li>
    </ul>
    <p>
        Los datos están diseñados para ser anónimos, utilizando identificadores únicos (UUID) en lugar de información personal identificable. Esto garantiza la privacidad de los postulantes al permitir un análisis en profundidad de las tendencias y características compartidas por los diferentes grupos de postulantes.
    </p>
</div>


# <a id="1"><font size="6"><div style="border-radius:5px;color:white;background-color:#FA8072;">  2. Comprensión del Negocio y de los terminos 📃</div> </font></a>
<a id="1"><font size="6">



<div style="border-radius:10px; border:#DEB887 solid; padding: 15px; background-color: #FFFAF0; font-size:100%; text-align:left;color: black">
<h3 align="left"><font color='#DEB887'>🌐 Introducción a la Detección de Comunidades:</font></h3>
<ul>
    <li><b>🤔 ¿Qué es la Detección de Comunidades?</b></li> 
    <ul>
        <li>La detección de comunidades es una técnica utilizada para identificar grupos o clústeres de nodos en una red que están más estrechamente conectados entre sí que con el resto de la red. Estas comunidades a menudo reflejan propiedades o funcionalidades compartidas.</li>
    </ul>
</ul>
<p>La detección de comunidades es fundamental en la ciencia de redes, ya que desentraña las agrupaciones intrínsecas dentro de redes complejas. Diferentes algoritmos y técnicas se adaptan a diversas estructuras de red y objetivos de detección, lo que hace crucial elegir un enfoque alineado con los datos y los objetivos generales.</p>
<div align="center" style="border: 2px solid #E57373; border-radius: 10px; padding: 20px; background-color: #FFCDD2;">
    <a href="https://stackoverflow.com/questions/23702543/d3-js-how-to-draw-a-force-directed-graph-with-nodes-grouped-by-circles" target="_blank">
        <img src="../images/l3RQc.png" alt="Red de Nodos" style="border-radius: 5px; max-width: 80%; box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);">
    </a>
    <p style="font-size: 16px; color: #0a0a0a; margin-top: 10px;">
        <em> <strong>Deteccion de Comunidades</strong>: en la imagen se observa colores diferentes en 
        <br>
        los nodos para indicar la comunidad diferente a la que pertenecen</em>
    </p>
</div>
</div>


<div style="border: 2px solid #2196F3; border-radius: 10px; padding: 20px; background-color: #E3F2FD;color: black">
<h3 style="color: #2196F3;">🧠 Definición de Grafo de Conocimiento</h3>
<p>Un <span style="color: #FF5722; font-weight: bold;">grafo de conocimiento</span> (KG) es un marco fundamental para representar entidades y las relaciones que las conectan. Los KG ofrecen una abstracción concisa e intuitiva para diversos dominios, donde los bordes capturan relaciones semánticas entre las entidades.</p>
<p>En términos matemáticos, un grafo de conocimiento se define como:</p>
<p align="center"><img src="https://latex.codecogs.com/svg.image?\dpi{120}&space;\large&space;KG&space;=&space;\{E,&space;R,&space;F\}" title="\large KG = \{E, R, F\}" /></p>
<ul>
  <li><strong>Entidades</strong> (<em>E</em>): Un conjunto de entidades presentes en el grafo.</li>
  <li><strong>Relaciones</strong> (<em>R</em>): Un conjunto de relaciones entre las entidades.</li>
  <li><strong>Hechos</strong> (<em>F</em>): Una colección de tripletas que denotan los enlaces entre las entidades a través de las relaciones.</li>
</ul>
<p>Un hecho se representa como una tripleta (<em>h, r, t</em>), donde <em>h</em> es la entidad principal (sujeto), <em>r</em> es la relación (predicado) y <em>t</em> es la entidad final (objeto).</p>
<p>Los KG son esenciales para la representación estructurada del conocimiento en una variedad de dominios y permiten inferir nuevos conocimientos a partir de los hechos presentes en el grafo.</p>
<div align="center" style="border: 2px solid #E57373; border-radius: 10px; padding: 20px; background-color: #E3F2FD;">
    <a href="[about:blank](https://www.analyticsvidhya.com/blog/2018/04/introduction-to-graph-theory-network-analysis-python-codes/)" target="_blank">
        <img src="../images/GNpython2.jpg" alt="Red de Nodos" style="border-radius: 5px; max-width: 80%; box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);">
    </a>
    <p style="font-size: 16px; color: #0a0a0a; margin-top: 10px;">
        <em> <strong>GC</strong>: Una imagen vale mas que mil palabras  
        <br>
        los nodos para indicar los postulantes y las aristas las relaciones entre ellos</em>
    </p>
</div>
</div>

https://en.wikipedia.org/wiki/Knowledge_graph_embedding

https://www.mdpi.com/2078-2489/13/4/161


<div style="border: 2px solid #4CAF50; border-radius: 10px; padding: 20px; background-color: #E8F5E9;color: black">
<h3 style="color: #4CAF50;">🌐 Embedding para Datos de Gráficos</h3>
<p>El <span style="color: #E91E63; font-weight: bold;">embedding</span> es una técnica que permite representar entidades y relaciones en un grafo en un espacio vectorial de baja dimensión. 
En el contexto de datos de gráficos, el embedding captura las características semánticas y topológicas de los nodos y las relaciones en el grafo, transformándolos en vectores numéricos.</p>
<p>Estos vectores permiten una manipulación y análisis más eficiente de los datos, facilitando tareas como la detección de similitudes, el clustering y la visualización.</p>
</div>
<div align="center" style="border: 2px solid #E57373; border-radius: 10px; padding: 20px; background-color: #E3F2FD;">
    <a href="[about:blank](https://www.analyticsvidhya.com/blog/2018/04/introduction-to-graph-theory-network-analysis-python-codes/)" target="_blank">
        <img src="../images/Embedding.webp" alt="Red de Nodos" style="border-radius: 5px; max-width: 80%; box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);">
    </a>
    <p style="font-size: 16px; color: #0a0a0a; margin-top: 10px;">
        <em> <strong>Embedding</strong>:
        Proceso de Incrustacion de Node2Vect</em>
    </p>
</div>
<div style="border: 2px solid #FF9800; border-radius: 10px; padding: 20px; background-color: #FFF3E0;color: black">
<h3 style="color: #FF9800;">📊 Node2Vec: Técnica de Embedding</h3>
<p><span style="color: #673AB7; font-weight: bold;">Node2Vec</span> es una técnica popular de embedding para grafos que permite generar representaciones vectoriales de nodos basadas en su estructura local en el grafo. Utiliza el concepto de "random walk" para capturar la proximidad semántica y topológica de los nodos.</p>
<p>Node2Vec utiliza un parámetro de control para ajustar el equilibrio entre explorar áreas distantes y explotar conexiones cercanas en el grafo. Los resultados son vectores numéricos que capturan la similitud y relación entre nodos en un espacio dimensional reducido.</p>
</div>
<div align="center" style="border: 2px solid #E57373; border-radius: 10px; padding: 20px; background-color: #E3F2FD;">
    <a href="[about:blank](https://www.analyticsvidhya.com/blog/2018/04/introduction-to-graph-theory-network-analysis-python-codes/)" target="_blank">
        <img src="../images/wor2vect.webp" alt="Red de Nodos" style="border-radius: 5px; max-width: 80%; box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);">
    </a>
    <p style="font-size: 16px; color: #0a0a0a; margin-top: 10px;">
        <em> <strong>Nod2Vect</strong>:
        Defincion matematica de Nod2Vect</em>
    </p>
</div>
<div style="border: 2px solid #2196F3; border-radius: 10px; padding: 20px; background-color: #E3F2FD;color: black">
<h3 style="color: #2196F3;">🔗 NetworkX en Python</h3>
<p><span style="color: #FF5722; font-weight: bold;">NetworkX</span> es una biblioteca de Python utilizada para la creación, manipulación y análisis de estructuras de grafos. Proporciona una amplia gama de funciones para crear, visualizar y analizar grafos, así como para realizar cálculos y aplicar algoritmos de grafos.</p>
<p>NetworkX es útil para procesar y preparar datos de grafos antes de aplicar técnicas como embedding. Permite cargar, modificar y explorar la estructura y atributos de los grafos, lo que es esencial para el proceso de embedding.</p>
</div>

<div style="border: 2px solid #FFC107; border-radius: 10px; padding: 20px; background-color: #FFF9C4;color: black">
<h3 style="color: #FFC107;">🚀 Proceso de Incrustación de Datos</h3>
<p>El proceso de <span style="color: #009688; font-weight: bold;">incrustación de datos</span> implica la transformación de los nodos y relaciones de un grafo en vectores numéricos de baja dimensión. Para llevar a cabo la incrustación de los postulantes en el proceso de admisión de la UNIQ, se seguirán los siguientes pasos:</p>
<ol>
  <li>Preparación de Datos: Cargar los datos de los postulantes y construir un grafo representando las relaciones entre ellos.</li>
  <li>Aplicación de Node2Vec: Utilizar Node2Vec para generar los vectores de embedding para cada postulante en el grafo. Ajustar los parámetros de exploración y explotación según sea necesario.</li>
  <li>Exploración y Análisis: Explorar las representaciones de embedding para identificar similitudes y diferencias entre los postulantes. Esto puede involucrar técnicas de clustering y visualización.</li>
  <li>Aplicación de Resultados: Utilizar los vectores de embedding en análisis posteriores, como clustering de postulantes, detección de tendencias y patrones, y toma de decisiones informadas en el proceso de admisión.</li>
</ol>

</div>


# <a id="1"><font size="6"><div style="border-radius:5px;color:white;background-color:#FA8072;"> 3. Comprensión de los datos 📚</div> </font></a>


Descripción del Conjunto de Datos: Tienes una descripción detallada de la información contenida en el conjunto de datos. Esto incluye información sobre los postulantes al proceso de admisión en la Universidad Nacional Intercultural de Quillabamba (UNIQ) en los años 2019 al 2023. También se mencionan las carreras profesionales, modalidades de postulación y otros detalles relevantes.


## <a><font size="6"><div style="border-radius:5px;color:#2F4F4F;background-color:#BFBF00;"> 🏋️‍♀️ 3.1 Librerias</div></font></a>


A continuacion se importan las librerias para analizar el conjutno de datos


In [92]:
# Importar la librería pandas para manejo de datos
import pandas as pd
import plotly.graph_objs as go
import plotly.offline as py
import seaborn as sns
import numpy as np

## <a><font size="6"><div style="border-radius:5px;color:#2F4F4F;background-color:#BFBF00;"> 🏋️‍♀️ 3.2 Lectura del conjunto de datos</div></font></a>


In [93]:
# Ruta al archivo CSV en tu Google Drive
csv_file_path = "../data/LISTA_DE_POSTULANTES_1.csv"

# Leer el archivo CSV en un DataFrame, especificando la codificación
Postulantes = pd.read_csv(csv_file_path, encoding="latin1", delimiter=";")

# Mostrar los primeros registros del DataFrame
Postulantes.head()

FECHA_CORTE                                               UUID  \
0     20230626  eb35ae9c84a9e239398331595b9f5489d23e686ca3f699...   
1     20230626  37aa167373b8ddd18910267b616ff845eef0517201f7da...   
2     20230626  41b4b33d0e004fc7dc6018664697814b2bfd87d93dafd9...   
3     20230626  60d7c1c310363c4b2bf1437aba8ba8df2a6f66cae4e25c...   
4     20230626  28a9a7dc2ae0a4c9d17151793585c9cee258418321d11b...   

                          ESCUELA  \
0                INGENIERÍA CIVIL   
1                      ECOTURISMO   
2                INGENIERÍA CIVIL   
3  INGENIERÍA AGRONÓMICA TROPICAL   
4         INGENIERÍA DE ALIMENTOS   

                                           MODALIDAD  ANNIO_POSTULACION  \
0  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
1  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
2  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
3  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
4  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   

   UBIGEO DEPARTAMENTO      PROVINCIA   DISTRITO  UBIGEO_ORIGEN  ... EDAD  \
0   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   20   
1   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   20   
2   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   19   
3   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   20   
4   80902        CUSCO  LA CONVENCIÓN   ECHARATE          80902  ...   20   

        SEXO PAGO   IDIOMA LEE ESCRIBE HABLA  TIENE_DISCAPACIDAD  \
0  MASCULINO  250  Español  SI      NO    NO                  NO   
1  MASCULINO  250  Español  SI      NO    NO                  NO   
2  MASCULINO  250  Español  SI      NO    NO                  NO   
3  MASCULINO  250  Español  SI      NO    NO                  NO   
4   FEMENINO  200  Español  SI      NO    NO                  NO   

  TIPO_COMUNIDAD  COMUNIDAD  
0        NINGUNA    NINGUNA  
1        NINGUNA    NINGUNA  
2        NINGUNA    NINGUNA  
3        NINGUNA    NINGUNA  
4        NINGUNA    NINGUNA  

[5 rows x 27 columns]

<div style="background-color: #f4f4f4; border: 1px solid #ddd; padding: 20px; border-radius: 5px; color: black">
    <h3 style="color: #2E86C1;">Simplificación de la Variable UUID en el Conjunto de Datos</h3>
    <p>Como <strong>científico de datos</strong>, se realizó la simplificación de la variable UUID en el conjunto de datos <code>Postulantes</code>. La variable UUID (Identificador Único Universal) se utiliza para identificar registros de manera única en un conjunto de datos. En este caso, se aplicó una transformación que conserva solo los primeros tres y últimos tres caracteres de cada UUID, reduciendo su longitud.</p>
    <p><strong>Ventajas:</strong></p>
    <ul>
        <li>Reducción de la longitud del UUID, facilitando la visualización y manipulación.</li>
        <li>Posible mejora en la eficiencia computacional al trabajar con datos extensos.</li>
    </ul>
    <p><strong>Consideraciones:</strong></p>
    <ul>
        <li>La simplificación puede llevar a la pérdida de información significativa.</li>
        <li>Se recomienda evaluar si la pérdida de información es aceptable y si el UUID simplificado sigue siendo único.</li>
    </ul>
    <p>La decisión de simplificar el UUID depende del contexto y el propósito del análisis. Se debe considerar si la pérdida de información es adecuada y si el UUID simplificado aún permite identificar registros de manera única.</p>
    <p style="color: #2E86C1;"><em>Recomendación:</em> Conservar el UUID original en su forma completa para análisis detallados o para enlaces con otras fuentes.</p>
    <p>Para más información, se puede consultar el código y la transformación realizada en el conjunto de datos.</p>
</div>


In [94]:
# Simplificar la variable UUID
Postulantes["UUID"] = Postulantes["UUID"].apply(lambda x: x[:3] + x[-3:])

# Mostrar el conjunto de datos con la variable UUID simplificada
Postulantes.head()

FECHA_CORTE    UUID                         ESCUELA  \
0     20230626  eb33a3                INGENIERÍA CIVIL   
1     20230626  37aa52                      ECOTURISMO   
2     20230626  41b65f                INGENIERÍA CIVIL   
3     20230626  60d267  INGENIERÍA AGRONÓMICA TROPICAL   
4     20230626  28acbc         INGENIERÍA DE ALIMENTOS   

                                           MODALIDAD  ANNIO_POSTULACION  \
0  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
1  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
2  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
3  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   
4  PRIMERA OPORTUNIDAD PARA ESTUDIANTES DEL QUINT...               2020   

   UBIGEO DEPARTAMENTO      PROVINCIA   DISTRITO  UBIGEO_ORIGEN  ... EDAD  \
0   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   20   
1   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   20   
2   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   19   
3   80901        CUSCO  LA CONVENCIÓN  SANTA ANA          80901  ...   20   
4   80902        CUSCO  LA CONVENCIÓN   ECHARATE          80902  ...   20   

        SEXO PAGO   IDIOMA LEE ESCRIBE HABLA  TIENE_DISCAPACIDAD  \
0  MASCULINO  250  Español  SI      NO    NO                  NO   
1  MASCULINO  250  Español  SI      NO    NO                  NO   
2  MASCULINO  250  Español  SI      NO    NO                  NO   
3  MASCULINO  250  Español  SI      NO    NO                  NO   
4   FEMENINO  200  Español  SI      NO    NO                  NO   

  TIPO_COMUNIDAD  COMUNIDAD  
0        NINGUNA    NINGUNA  
1        NINGUNA    NINGUNA  
2        NINGUNA    NINGUNA  
3        NINGUNA    NINGUNA  
4        NINGUNA    NINGUNA  

[5 rows x 27 columns]

In [95]:
Postulantes["UBIGEO"].value_counts()

80901     1813
80909      567
80902      480
80914      291
80906      195
          ... 
150106       1
40501        1
80708        1
200101       1
100101       1
Name: UBIGEO, Length: 213, dtype: int64

### Gráfico de Histograma del conjunto de datos Postulantes


In [96]:
!pip install dash

In [97]:
# Importar bibliotecas
from dash import Dash, html, dcc, callback, Output, Input
import pandas as pd
import plotly.express as px

# Inicializar la aplicación
app = Dash(__name__)

# Diseño de la aplicación
app.layout = html.Div(
    style={"textAlign": "center", "padding": "20px"},
    children=[
        html.H1("Gráfico de Histograma Interactivo", style={"color": "#2E86C1"}),
        html.Hr(),
        dcc.Dropdown(
            id="dropdown-column",
            options=[{"label": col, "value": col} for col in Postulantes.columns],
            value="EDAD",
            multi=False,
            style={"width": "50%", "margin": "auto"},
        ),
        dcc.Graph(id="histogram", style={"margin-top": "20px"}),
    ],
)


# Definir la función de devolución de llamada
@callback(
    Output(component_id="histogram", component_property="figure"),
    Input(component_id="dropdown-column", component_property="value"),
)
def update_histogram(selected_column):
    fig = px.histogram(
        Postulantes,
        x=selected_column,
        title=f"Histograma de {selected_column}",
        color_discrete_sequence=px.colors.sequential.Viridis,
    )
    fig.update_layout(
        plot_bgcolor="#F9F9F9",  # Color de fondo del gráfico
        paper_bgcolor="#F9F9F9",  # Color de fondo del área de trazado
        font_color="#333333",  # Color de fuente
    )
    return fig


app.run_server(debug=True)

Address already in use
Port 8050 is in use by another program. Either identify and stop that program, or start the server with a different port.


AttributeError: 'tuple' object has no attribute 'tb_frame'

## <a><font size="6"><div style="border-radius:5px;color:#2F4F4F;background-color:#BFBF00;"> 🏋️‍♀️ 3.3 Acerca del Conjunto de Datos</div></font></a>


<div style="background-color:#20B2AA; border-radius:10px; padding:10px; box-shadow: 0px 2px 4px rgba(10, 39, 52, 0.1);">
    <h3 style="color:#333333;">📊 - Tamaño del conjunto de datos</h3>
    <p style="color:#555555;">Este es el tamaño actual del conjunto de datos:</p>
    <p style="color:#333333;">- <strong>Tamaño del conjunto de datos:</strong> (4928, 27)</p>
</div>


In [ ]:
# Verificar el número de filas y columnas
num_filas, num_columnas = Postulantes.shape
print(f"Número de filas: {num_filas}")
print(f"Número de columnas: {num_columnas}")

**Observaciones:**

- El conjunto de datos "Postulantes" tiene un total de 4928 filas y 27 columnas.
- Cada fila representa un postulante y
- cada columna contiene diferentes atributos y características relacionadas con los postulantes


<div style="background-color:#20B2AA; border-radius:10px; padding:10px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h3 style="color:#333333;">📊 Tipología del conjunto de datos</h3>
    <p style="color:#555555;">Esta sección muestra la tipología del conjunto de datos:</p>
</div>


In [ ]:
# Tipos de datos de cada columna
print(Postulantes.dtypes)

<style>
    table {
        border-collapse: collapse;
        width: 100%;
        border: 2px solid #008CBA;
        margin-bottom: 20px;
    }

    th, td {
        border: 1px solid #008CBA;
        padding: 10px;
        text-align: left;
    }

    th {
        background-color: #008CBA;
        color: white;
    }

    h2 {
        color: #008CBA;
    }
</style>

#### <span style="color:#008CBA;">Variables Numéricas (`int64`):</span>

<table>
    <tr>
        <th>Variable</th>
        <th>Descripción</th>
    </tr>
    <tr>
        <td>FECHA_CORTE</td>
        <td>Día en que se generó el DATASET</td>
    </tr>
    <tr>
        <td>ANNIO_POSTULACION</td>
        <td>Año del proceso de admisión</td>
    </tr>
    <tr>
        <td>UBIGEO</td>
        <td>Código de ubicación geográfica donde nació el postulante</td>
    </tr>
    <tr>
        <td>UBIGEO_ORIGEN</td>
        <td>Código de ubicación geográfica donde vive el postulante</td>
    </tr>
    <tr>
        <td>UBIGEO_IE</td>
        <td>Código de ubicación geográfica donde culminó sus estudios el postulante</td>
    </tr>
    <tr>
        <td>EDAD</td>
        <td>Número que identifica la edad del postulante en la fecha que postuló</td>
    </tr>
    <tr>
        <td>PAGO</td>
        <td>Monto del pago que realizó el postulante para dar el examen de admisión</td>
    </tr>
</table>

#### <span style="color:#008CBA;">Variables de Categorías (`object`):</span>

<table>
    <tr>
        <th>Variable</th>
        <th>Descripción</th>
    </tr>
    <tr>
        <td>UUID</td>
        <td>Documento Nacional de Identidad del alumno que postuló, anonimizado</td>
    </tr>
    <tr>
        <td>ESCUELA</td>
        <td>Escuela Profesional a la que postuló el alumno</td>
    </tr>
    <tr>
        <td>MODALIDAD</td>
        <td>Modalidad de Postulación</td>
    </tr>
    <tr>
        <td>DEPARTAMENTO</td>
        <td>Nombre del departamento donde nació el postulante</td>
    </tr>
    <tr>
        <td>PROVINCIA</td>
        <td>Nombre de la provincia donde nació el postulante</td>
    </tr>
    <tr>
        <td>DISTRITO</td>
        <td>Nombre del distrito donde nació el postulante</td>
    </tr>
    <tr>
        <td>DEPARTAMENTO_ORIGEN</td>
        <td>Nombre del departamento donde vive el postulante</td>
    </tr>
    <tr>
        <td>PROVINCIA_ORIGEN</td>
        <td>Nombre de la provincia donde vive el postulante</td>
    </tr>
    <tr>
        <td>DISTRITO_ORIGEN</td>
        <td>Nombre del distrito donde vive el postulante</td>
    </tr>
    <tr>
        <td>DEPARTAMENTO_IE</td>
        <td>Nombre del departamento donde culminó sus estudios el postulante</td>
    </tr>
    <tr>
        <td>PROVINCIA_IE</td>
        <td>Nombre de la provincia donde culminó sus estudios el postulante</td>
    </tr>
    <tr>
        <td>DISTRITO_IE</td>
        <td>Nombre del distrito donde culminó sus estudios el postulante</td>
    </tr>
    <tr>
        <td>SEXO</td>
        <td>Descripción del sexo del postulante</td>
    </tr>
    <tr>
        <td>IDIOMA</td>
        <td>Segunda lengua del estudiante</td>
    </tr>
    <tr>
        <td>LEE</td>
        <td>Lee la segunda lengua</td>
    </tr>
    <tr>
        <td>ESCRIBE</td>
        <td>Escribe la segunda lengua</td>
    </tr>
    <tr>
        <td>HABLA</td>
        <td>Habla la segunda lengua</td>
    </tr>
    <tr>
        <td>TIENE_DISCAPACIDAD</td>
        <td>Si el postulante tiene alguna discapacidad</td>
    </tr>
    <tr>
        <td>TIPO_COMUNIDAD</td>
        <td>Tipo de comunidad a la cual pertenece el postulante</td>
    </tr>
    <tr>
        <td>COMUNIDAD</td>
        <td>Comunidad específica a la cual pertenece el postulante</td>
    </tr>
</table>


<div style="background-color:#20B2AA; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h3 style="color:#191970;">📊 Descripción del conjunto de datos</h3>
    <p style="color:#555555;">Esta sección proporciona una descripción del conjunto de datos:</p>
</div>


In [ ]:
# Resumen estadístico de las variables numericas
Postulantes.describe()

<style>
    table {
        border-collapse: collapse;
        width: 70%;
        border: 2px solid #008CBA;
        margin-bottom: 20px;
    }

    th, td {
        border: 1px solid #008CBA;
        padding: 10px;
        text-align: center;
    }

    th {
        background-color: #008CBA;
        color: white;
    }

    h2 {
        color: #008CBA;
    }

    .stats-table {
        margin: 0 auto;
    }
</style>

#### <span style="color:#008CBA;">Resumen Estadístico</span>

<div class="stats-table">
    <table>
        <tr>
            <th>Variable</th>
            <th>count</th>
            <th>mean</th>
            <th>std</th>
            <th>min</th>
            <th>25%</th>
            <th>50%</th>
            <th>75%</th>
            <th>max</th>
        </tr>
        <tr>
            <td>FECHA_CORTE</td>
            <td>4928.0</td>
            <td>20230626.0</td>
            <td>0.0</td>
            <td>20230626.0</td>
            <td>20230626.0</td>
            <td>20230626.0</td>
            <td>20230626.0</td>
            <td>20230626.0</td>
        </tr>
        <tr>
            <td>ANNIO_POSTULACION</td>
            <td>4928.00000</td>
            <td>2021.13332</td>
            <td>1.49930</td>
            <td>2019.00000</td>
            <td>2020.00000</td>
            <td>2021.00000</td>
            <td>2023.00000</td>
            <td>2023.00000</td>
        </tr>
        <tr>
            <td>UBIGEO</td>
            <td>4928.000000</td>
            <td>83352.578531</td>
            <td>22286.639332</td>
            <td>20101.000000</td>
            <td>80901.000000</td>
            <td>80901.000000</td>
            <td>80909.000000</td>
            <td>250202.000000</td>
        </tr>
        <tr>
            <td>UBIGEO_ORIGEN</td>
            <td>4928.000000</td>
            <td>82377.273539</td>
            <td>17009.440949</td>
            <td>30101.000000</td>
            <td>80901.000000</td>
            <td>80902.000000</td>
            <td>80909.000000</td>
            <td>250202.000000</td>
        </tr>
        <tr>
            <td>UBIGEO_IE</td>
            <td>4928.000000</td>
            <td>83585.048295</td>
            <td>21914.275166</td>
            <td>20401.000000</td>
            <td>80901.000000</td>
            <td>80901.000000</td>
            <td>80909.000000</td>
            <td>250202.000000</td>
        </tr>
        <tr>
            <td>EDAD</td>
            <td>4928.000000</td>
            <td>20.236810</td>
            <td>4.162145</td>
            <td>15.000000</td>
            <td>18.000000</td>
            <td>19.000000</td>
            <td>21.000000</td>
            <td>67.000000</td>
        </tr>
        <tr>
            <td>PAGO</td>
            <td>4928.000000</td>
            <td>179.135552</td>
            <td>64.030636</td>
            <td>0.000000</td>
            <td>200.000000</td>
            <td>200.000000</td>
            <td>200.000000</td>
            <td>670.000000</td>
        </tr>
    </table>
</div>


<div style="background-color: #f4f4f4; padding: 20px; border: 1px solid #dcdcdc; border-radius: 5px; font-family: Arial, sans-serif; color: #333;">
  <h3 style="color: #1f77b4;">Observacion de Variables Numéricas:</h3>
  
  <ul>
    <li><strong>FECHA_CORTE:</strong> La columna FECHA_CORTE tiene un valor constante de 20230626, lo que indica que todos los datos en esta columna son del mismo día. &#x1F4C5;</li>
    <li><strong>ANNIO_POSTULACION:</strong> El año promedio de proceso de admisión es aproximadamente 2021, con un rango entre 2019 y 2023. &#x1F4C5;</li>
    <li><strong>UBIGEO:</strong> Los códigos de ubicación geográfica donde nacieron los postulantes varían significativamente, con un promedio alrededor de 83352.578531. &#x1F4C5;</li>
    <li><strong>UBIGEO_ORIGEN:</strong> Los códigos de ubicación geográfica donde viven los postulantes también varían, con un promedio alrededor de 82377.273539. &#x1F4C5;</li>
    <li><strong>UBIGEO_IE:</strong> Los códigos de ubicación geográfica donde culminaron sus estudios los postulantes varían, con un promedio alrededor de 83585.048295. &#x1F4C5;</li>
    <li><strong>EDAD:</strong> La edad promedio de los postulantes es de aproximadamente 20 años, con un rango entre 15 y 67 años. &#x1F4C5;</li>
    <li><strong>PAGO:</strong> El monto promedio de pago que realizaron los postulantes para dar el examen de admisión es de alrededor de 179.135552, con un rango amplio de valores. &#x1F4C5;</li>
  </ul>
</div>


In [ ]:
# Verificar resumen estadístico para variables categóricas
resumen_categoricas = Postulantes.describe(include=["object"])

# Mostrar el resumen estadístico
resumen_categoricas

<div style="background-color: #f4f4f4; padding: 20px; border: 1px solid #dcdcdc; border-radius: 5px; font-family: Arial, sans-serif; color: #333;">
  <h3 style="color: #1f77b4;">Observacion de Variables Categóricas :</h3>
  
  <ul>
    <li><strong>UUID:</strong> Hay 3748 valores únicos en esta columna. El valor más común (modo) aparece 6 veces. &#x1F4C3;</li>
    <li><strong>ESCUELA:</strong> Hay 6 valores únicos en esta columna. La escuela "INGENIERÍA CIVIL" es la más común con 1813 apariciones. &#x1F4C3;</li>
    <li><strong>MODALIDAD:</strong> Hay 8 valores únicos en esta columna. La modalidad "ORDINARIO" es la más común con 2844 apariciones. &#x1F4C3;</li>
    <li><strong>DEPARTAMENTO:</strong> Hay 21 valores únicos en esta columna. El departamento "CUSCO" es el más común con 4483 apariciones. &#x1F4C3;</li>
    <li><strong>PROVINCIA:</strong> Hay 65 valores únicos en esta columna. La provincia "LA CONVENCIÓN" es la más común con 3893 apariciones. &#x1F4C3;</li>
    <li><strong>DISTRITO:</strong> Hay 204 valores únicos en esta columna. El distrito "SANTA ANA" es el más común con 1813 apariciones. &#x1F4C3;</li>
    <li><strong>DEPARTAMENTO_ORIGEN:</strong> Hay 12 valores únicos en esta columna. El departamento "CUSCO" es el más común con 4742 apariciones. &#x1F4C3;</li>
    <li><strong>PROVINCIA_ORIGEN:</strong> Hay 41 valores únicos en esta columna. La provincia "LA CONVENCIÓN" es la más común con 4404 apariciones. &#x1F4C3;</li>
    <li><strong>DISTRITO_ORIGEN:</strong> Hay 142 valores únicos en esta columna. El distrito "SANTA ANA" es el más común con 2008 apariciones. &#x1F4C3;</li>
    <li><strong>DEPARTAMENTO_IE:</strong> Hay 15 valores únicos en esta columna. El departamento "CUSCO" es el más común con 4609 apariciones. &#x1F4C3;</li>
    <li><strong>PROVINCIA_IE:</strong> Hay 46 valores únicos en esta columna. La provincia "LA CONVENCIÓN" es la más común con 4057 apariciones. &#x1F4C3;</li>
    <li><strong>DISTRITO_IE:</strong> Hay 150 valores únicos en esta columna. El distrito "SANTA ANA" es el más común con 1921 apariciones. &#x1F4C3;</li>
    <li><strong>SEXO:</strong> Hay 2 valores únicos en esta columna. "MASCULINO" es el más común con 2860 apariciones. &#x1F4C3;</li>
    <li><strong>IDIOMA:</strong> Hay 9 valores únicos en esta columna. "Español" es el más común con 3832 apariciones. &#x1F4C3;</li>
    <li><strong>LEE:</strong> Hay 2 valores únicos en esta columna. "SI" es el más común con 2844 apariciones. &#x1F4C3;</li>
    <li><strong>ESCRIBE:</strong> Hay 2 valores únicos en esta columna. "SI" es el más común con 2944 apariciones. &#x1F4C3;</li>
    <li><strong>HABLA:</strong> Hay 2 valores únicos en esta columna. "SI" es el más común con 3913 apariciones. &#x1F4C3;</li>
    <li><strong>TIENE_DISCAPACIDAD:</strong> Hay 2 valores únicos en esta columna. "NO" es el más común con 3619 apariciones. &#x1F4C3;</li>
    <li><strong>TIPO_COMUNIDAD:</strong> Hay 3 valores únicos en esta columna. "NINGUNA" es la más común con 3619 apariciones. &#x1F4C3;</li>
    <li><strong>COMUNIDAD:</strong> Hay 5 valores únicos en esta columna. "NINGUNA" es la más común con 3619 apariciones. &#x1F4C3;</li>
  </ul>
</div>


<div style="background-color:#48D1CC; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h2 style="color:#C71585;">🧮 Verificación de valores nulos</h2>
    <p style="color:#555555;">Esta sección muestra la cantidad de valores nulos en el conjunto de datos:</p>
</div>


In [ ]:
# Cantidad de valores nulos
print(Postulantes.isnull().sum())

<div style="background-color: #f5f5f5; border: 1px solid #dddddd; padding: 10px; border-radius: 5px; font-family: Arial, sans-serif; color: #333333;">
    <p style="font-size: 16px; font-weight: bold; color: #ff9900;">🚫 Valores Nulos en el Conjunto de Datos 🚫</p>
    <p>La columna '<span style="color: #ff0000;">IDIOMA</span>' tiene <span style="color: #ff0000;">984</span> valores nulos. Esto significa que hay <span style="color: #ff0000;">984</span> registros en los cuales no se proporcionó información sobre la segunda lengua del estudiante.</p>
    <p>Las columnas '<span style="color: #ff0000;">LEE</span>', '<span style="color: #ff0000;">ESCRIBE</span>', '<span style="color: #ff0000;">HABLA</span>', '<span style="color: #ff0000;">TIENE_DISCAPACIDAD</span>', '<span style="color: #ff0000;">TIPO_COMUNIDAD</span>' y '<span style="color: #ff0000;">COMUNIDAD</span>' también tienen <span style="color: #ff0000;">984</span> valores nulos cada una. Esto sugiere que en estos registros, no se proporcionó información sobre si el estudiante lee, escribe, habla una segunda lengua, tiene discapacidad, o sobre su tipo de comunidad y comunidad específica.</p>
    En conclusion 
</div>


<div style="background-color:#48D1CC; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h3 style="color:#191970;">📊 Porcentaje y Ubicación de valores nulos</h3>
    <p style="color:#555555;">En esta sección se presenta el porcentaje de valores nulos por columna y una representación gráfica de la ubicación de los valores faltantes:</p>
    <div style="display:flex; flex-direction:row; justify-content:space-between;">
        <div style="width:45%;">
            <p style="color:#333333;"><strong>Porcentaje de valores nulos:</strong></p>
            <ul style="color:#333333; padding-left:20px;">
                <li><strong>Columna 'IDIOMA':</strong> 984 valores nulos (20%)</li>
                <li><strong>Columna 'LEE':</strong> 984 valores nulos (20%)</li>
                <li><strong>Columna 'ESCRIBE':</strong> 984 valores nulos (20%)</li>
                <li><strong>Columna 'HABLA':</strong> 984 valores nulos (20%)</li>
                <li><strong>Columna 'TIENE_DISCAPACIDAD':</strong> 984 valores nulos (20%)</li>
                <li><strong>Columna 'TIPO_COMUNIDAD':</strong> 984 valores nulos (20%)</li>
                <li><strong>Columna 'COMUNIDAD':</strong> 984 valores nulos (20%)</li>
            </ul>
        </div>
        <div style="width:50%;">
            <p style="color:#333333;"><strong>Ubicación de valores nulos:</strong></p>
            <img src="../images/newplot.png" alt="Gráfico de Ubicación de Valores Nulos" style="max-width:100%; border: 1px solid #ddd;">
        </div>
    </div>
</div>


In [ ]:
import plotly.express as px
import pandas as pd

# Crear un DataFrame con los porcentajes de valores nulos por columna
null_percentages = (Postulantes.isnull().sum() / len(Postulantes)) * 100
null_df = pd.DataFrame(
    {
        "Columna": null_percentages.index,
        "Porcentaje de Valores Nulos": null_percentages.values,
    }
)

# Gráfico de barras para el porcentaje de valores nulos por columna
bar_fig = px.bar(
    null_df,
    x="Columna",
    y="Porcentaje de Valores Nulos",
    title="Porcentaje de Valores Nulos por Columna",
    labels={"Porcentaje de Valores Nulos": "Porcentaje"},
    template="plotly",
    color_discrete_sequence=["#1F77B4"],
)
bar_fig.update_layout(xaxis_title="Columnas", yaxis_title="Porcentaje")
bar_fig.update_xaxes(tickangle=-45, tickfont=dict(size=10))

# Gráfico de ubicación de valores faltantes
missing_data_fig = px.imshow(
    Postulantes.isnull(),
    labels=dict(x="Columnas", y="Filas"),
    title="Ubicación de Valores Faltantes",
    color_continuous_scale="Plasma",
    template="plotly",
    height=400,
)
missing_data_fig.update_layout(xaxis_title="Columnas", yaxis_title="Filas")

# Mostrar los gráficos
bar_fig.show()
missing_data_fig.show()

<div style="border: 2px solid #4CAF50; background-color: #E7F6E7; padding: 10px; border-radius: 5px; font-family: Arial, sans-serif; color: #333;">
    <p style="font-size: 18px; font-weight: bold;">🔍 Conclusión:</p>
    <p>Dado que estas columnas están relacionadas con el dominio del segundo idioma y la comunidad del postulante, aquí hay algunas recomendaciones, Ademas casi el 20% del conjunto de datos en las columnas de IDIOMA, LEE, ESCRIBE ... estan vacios, se sugiere:</p>
    <ul style="list-style-type: disc; margin-left: 20px;">
        <li>Imputar estos valores nulos con la moda o mediana de cada columna (opción más simple).</li>
        <li>Reemplazar los valores nulos por una categoría "No informado" en cada columna.</li>
        <li>Si hay un patrón en los nulos (por ejemplo, mismos postulantes), se podría intentar predecir los valores en base a otras columnas.</li>
    </ul>
</div>


<div style="background-color:#48D1CC; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h2 style="color:#000080;">📊 Datos Únicos</h2>
    <p style="color:#555555;">En esta sección se presenta la información sobre la cantidad de valores únicos en cada columna del conjunto de datos:</p>
</div>


In [ ]:
# Cantidad de valores únicos para cada columna
print(Postulantes.nunique())

<div style="background-color: #f5f5f5; border: 1px solid #ddd; padding: 15px; border-radius: 10px;color: black">
    <p style="font-size: 18px; color: #2E86C1;">Análisis de la Cantidad de Valores Únicos por Columna</p>
    <p>El análisis de la cantidad de valores únicos en cada columna proporciona una visión valiosa sobre la diversidad y variabilidad en el conjunto de datos <strong>Postulantes</strong>:</p>    
    <ul>
        <li>Hay 1 <strong>Fecha_Corte</strong>, lo que sugiere eliminar esta variable ya que no aporta información significativa.</li>
        <li>Hay 3748 <strong>UUIDs</strong> únicos, lo que indica múltiples identificadores de postulantes.</li>
        <li>Las columnas de <strong>ubicación geográfica</strong> tienen diferentes cantidades de valores únicos, reflejando la diversidad de ubicaciones de origen y educación.</li>
        <li>Existen <strong>varias escuelas profesionales y modalidades</strong> de postulación.</li>
        <li>Los años de <strong>postulación</strong> presentan 5 valores únicos, lo que podría indicar distintos años de procesos.</li>
        <li>La <strong>edad</strong> de los postulantes varía en 37 valores únicos, señalando diversidad de edades.</li>
        <li>El <strong>género</strong> de los postulantes es predominantemente binario.</li>
        <li>Las habilidades, discapacidades y pertenencia a <strong>comunidades</strong> tienen opciones binarias y algunas categorías.</li>
        <li>Las columnas <strong>IDIOMA</strong> y <strong>PAGO</strong> presentan múltiples opciones.</li>
    </ul> 
    <p>Este análisis es crucial para comprender la distribución y naturaleza de los datos, lo que guía las decisiones en el procesamiento y análisis subsiguiente.</p>
</div>


### Eliminar la columna 'Fecha_Corte' del conjutno de datos Postulantes


In [ ]:
# Eliminar la columna 'Fecha_Corte'
Postulantes = Postulantes.drop(columns=['FECHA_CORTE'])

# Mostrar el conjunto de datos después de eliminar la columna
Postulantes.head()


### Grafico Circular del conjunto de datos Postulanets


In [ ]:
from dash import Dash, html, dcc, callback, Output, Input
import pandas as pd
import plotly.express as px

# Inicializar la aplicación
app = Dash(__name__)


# Excluir la columna 'UUID'
excluded_columns = ["UUID"]
columns_for_dropdown = [
    col for col in Postulantes.columns if col not in excluded_columns
]

# Diseño de la aplicación
app.layout = html.Div(
    style={"textAlign": "center", "padding": "20px"},
    children=[
        html.H1("Gráfico Circular Interactivo", style={"color": "#2E86C1"}),
        html.Hr(),
        dcc.Dropdown(
            id="dropdown-column",
            options=[{"label": col, "value": col} for col in columns_for_dropdown],
            value="ESCUELA",
            multi=False,
            style={"width": "50%", "margin": "auto"},
        ),
        dcc.Graph(id="pie-chart", style={"margin-top": "20px"}),
    ],
)


# Definir la función de devolución de llamada
@callback(
    Output(component_id="pie-chart", component_property="figure"),
    Input(component_id="dropdown-column", component_property="value"),
)
def update_pie_chart(selected_column):
    pie_data = Postulantes[selected_column].value_counts()
    fig = px.pie(
        pie_data,
        names=pie_data.index,
        values=pie_data.values,
        title=f"Distribución de {selected_column}",
        hole=0.3,
    )
    fig.update_layout(
        plot_bgcolor="#F9F9F9",  # Color de fondo del gráfico
        paper_bgcolor="#F9F9F9",  # Color de fondo del área de trazado
        font_color="#333333",  # Color de fuente
    )
    return fig


app.run_server(debug=True)

<div style="background-color:#48D1CC; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h2 style="color:#000080;">📋 Datos Duplicados</h2>
    <p style="color:#555555;">En esta sección se analiza la presencia de datos duplicados en el conjunto de datos:</p>
    <p style="color:#555555;"><strong>Total de registros duplicados:</strong> 2075</p>
    <p style="color:#555555;">La duplicación de UUIDs podría indicar que algunos postulantes realizaron múltiples intentos de admisión.</p>
</div>

<div style="background-color:#f2f2f2; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h3 style="color:#2E86C1;">Verificación de Valores Duplicados</h3>
    <p style="color:#555555;">En esta sección se verifica la presencia de valores nulos en el dataset 'Postulantes':</p>
    <pre style="color:#555555; background-color:#ffffff; padding:10px; border:1px solid #dddddd; border-radius:5px;">
    Cantidad de filas duplicadas: 0 </pre>
</div>


In [ ]:
duplicates_count = Postulantes.duplicated().sum()
print("Cantidad de filas duplicadas:", duplicates_count)

<div style="background-color:#f2f2f2; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h3 style="color:#2E86C1;">Observación</h3>
    <ul style="color:#555555; list-style-type:square; padding-left:20px;">
        <li>No se observan datos faltantes en el dataset de <strong>Postulantes</strong>.</li>
        <li>Sin embargo,</li>
        <li>La variable <strong>UUID</strong> tiene 3748 registros únicos, sin embargo debería tener 4928 (Total de filas del dataset).</li>
        <li>Vamos a verificar cuáles son los datos que se duplican y por qué se duplican.</li>
    </ul>
</div>


<div style="background-color:#f2f2f2; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h3 style="color:#2E86C1;">Verificamos datos faltantes de la variable UUIDs
</h3>
</div>


### Verificamos datos faltantes de la variable UUIDs


In [ ]:
duplicated_uuids = Postulantes[Postulantes["UUID"].duplicated(keep=False)]
duplicated_uuid_counts = duplicated_uuids["UUID"].value_counts()

print("Registros duplicados en la variable UUID:")

print("Cantidad de filas duplicadas:", duplicated_uuids.shape[0])

<style>
  .observation-box {
    border: 2px solid #3498db;
    background-color: #f0f0f0;
    padding: 10px;
    margin-bottom: 15px;
    font-family: Arial, sans-serif;
    color: black
  }
  .emoji {
    font-size: 18px;
    margin-right: 5px;
  }
</style>

<div class="observation-box">
  <p><span class="emoji">🔍</span> Observación</p>
  <ul>
    <li>En total, se identificaron <strong>2075 filas duplicadas</strong> en la columna 'UUID'.</li>
    <li>La duplicación de UUIDs podría indicar que algunos postulantes realizaron <strong>múltiples intentos de admisión</strong>.</li>
    <li>Para confirmar esta hipótesis se analizó más a fondo los datos relacionados con los postulantes duplicados.</li>
  </ul>
</div>


### Grafico de Correlacion de datos de la columna UUID


In [ ]:
from dash import Dash, html, dcc, callback, Output, Input
import pandas as pd
import plotly.express as px

# Inicializar la aplicación
app = Dash(__name__)

# Excluir la columna 'UUID'
excluded_columns = ["UUID"]
columns_for_dropdown = [
    col for col in duplicated_uuids.columns if col not in excluded_columns
]

# Diseño de la aplicación
app.layout = html.Div(
    style={"textAlign": "center", "padding": "20px"},
    children=[
        html.H1("Gráfico de Correlación Interactivo", style={"color": "#2E86C1"}),
        html.Hr(),
        dcc.Dropdown(
            id="dropdown-column",
            options=[{"label": col, "value": col} for col in columns_for_dropdown],
            value="ESCUELA",
            multi=False,
            style={"width": "50%", "margin": "auto"},
        ),
        dcc.Graph(id="correlation-plot", style={"margin-top": "20px"}),
    ],
)


# Definir la función de devolución de llamada
@callback(
    Output(component_id="correlation-plot", component_property="figure"),
    Input(component_id="dropdown-column", component_property="value"),
)
def update_correlation_plot(selected_column):
    corr_data = duplicated_uuids[["UUID", selected_column]].head(20)
    fig = px.scatter(
        corr_data,
        x="UUID",
        y=selected_column,
        title=f"Correlación entre UUID y {selected_column}",
    )
    fig.update_layout(
        plot_bgcolor="#F9F9F9",  # Color de fondo del gráfico
        paper_bgcolor="#F9F9F9",  # Color de fondo del área de trazado
        font_color="#333333",  # Color de fuente
    )
    return fig


# Ejecutar la aplicación
app.run_server(debug=True)

In [ ]:
# Filtrar el DataFrame por el postulante con código UUID deseado
filtered_data = Postulantes[Postulantes["UUID"] == "90a0f9"]

# Mostrar todas las filas filtradas
filtered_data

<style>
  .analysis-container {
    background-color: #f4f4f4;
    border: 1px solid #ccc;
    padding: 20px;
    margin: 20px 0;
    border-radius: 5px;
    color: black;
  }

  .highlight {
    background-color: #ffc107;
    padding: 2px;
    border-radius: 1px;
  }
</style>

<div class="analysis-container">
  <h3>Análisis de Duplicación de Postulante</h3>
  <p>El resultado indica que el postulante con el código UUID "90a0f9" ha realizado dos postulaciones:</p>
  <ol>
    <li>Primera Postulación:</li>
    <ul>
      <li>Programa: <span class="highlight">Ingeniería Civil</span></li>
      <li>Modalidad: <span class="highlight">Primera Oportunidad para Estudiantes del Quinto de Secundaria</span></li>
      <li>Año de Postulación: <span class="highlight">2020</span></li>
      <li>Ubicación: <span class="highlight">Cusco, La Convención, Santa Ana</span></li>
    </ul>
    <li>Segunda Postulación:</li>
    <ul>
      <li>Programa: <span class="highlight">Ingeniería Agronómica Tropical</span></li>
      <li>Modalidad: <span class="highlight">Ordinario</span></li>
      <li>Año de Postulación: <span class="highlight">2020</span></li>
      <li>Ubicación: <span class="highlight">Cusco, La Convención, Santa Ana</span></li>
    </ul>
  </ol>
  <p>Esta duplicación de postulaciones podría sugerir que el postulante intentó ingresar a diferentes programas utilizando distintas modalidades de postulación en el mismo año.</p>
</div>


<style>
  .note-container {
    background-color: #f0f8ff;
    padding: 20px;
    border-radius: 10px;
    border: 1px solid #a3c2c2;
    color: black;
  }
  .conclusion {
    background-color: #e6f7ff;
    padding: 20px;
    border-radius: 10px;
    border: 1px solid #a3c2c2;
    color: black;
  }
</style>

<div class="note-container">
  <h3>📝 Análisis de Datos Duplicados en el Conjunto de Datos 'Postulantes'</h3>
  <p>En base a los datos proporcionados, se identificaron registros duplicados en la columna 'UUID' del conjunto de datos 'Postulantes'. Esto sugiere que algunos postulantes realizaron múltiples intentos de admisión, lo cual es una hipótesis plausible dada la naturaleza de los procesos de admisión y las oportunidades para mejorar los resultados.</p>
</div>

<div class="conclusion">
  <h3>🔍 Conclusiones</h3>
  <p>Después de un análisis más detenido, la decisión tomada es eliminar los datos duplicados mientras se conserva la última postulación realizada por cada postulante. Esto implica que se retendrá únicamente la última fila para cada UUID, lo que representa el último intento registrado por cada individuo.</p>
  <p>Esta acción tiene sentido para mantener la integridad y la precisión de los datos. Conservar solo la última postulación ayudará a evitar posibles incoherencias y redundancias en el conjunto de datos.</p>
</div>


In [ ]:
# Eliminar duplicados conservando la última postulación de cada postulante
postulantes_unicos = Postulantes.drop_duplicates(subset="UUID", keep="last")

# Mostrar el tamaño del conjunto de datos original 'Postulantes'
print("Tamaño del conjunto de datos original 'Postulantes':", Postulantes.shape)

# Mostrar el tamaño del conjunto de datos 'postulantes_sin_duplicados'
print("Tamaño del conjunto de datos 'postulantes_unicos':", postulantes_unicos.shape)

<div style="background-color:#f7f7f7; border-radius:10px; padding:20px;color:black">

### Análisis de Datos - Eliminación de Duplicados

En base a los datos duplicados en el conjunto de datos de los postulantes, se realizó un proceso de limpieza para conservar únicamente la última postulación realizada por cada postulante. Esto se logró mediante la eliminación de duplicados y la retención de la última fila correspondiente a cada UUID.

**Tamaño del conjunto de datos original 'Postulantes':** (4928, 27)

**Tamaño del conjunto de datos 'postulantes_unicos' después de eliminar duplicados:** (3748, 27)

El objetivo de esta acción es asegurarse de que solo se conserve la información más reciente de cada postulante, eliminando registros duplicados que podrían haber sido causados por múltiples intentos de admisión.

</div>


<div style="background-color:#48D1CC; border-radius:10px; padding:15px; box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);">
    <h2 style="color:#000080;">📋 Convercion de tipos de datos</h2>
    <p style="color:#555555;">En esta sección se convertiran los datos al formato adecuado</p>
</div>

In [ ]:
# Convertir la columna PAGO al tipo de dato de flotante
postulantes_unicos.loc[:, "PAGO"] = postulantes_unicos["PAGO"].astype(float)

In [ ]:
postulantes_unicos.head()

In [ ]:
filas = len(postulantes_unicos.columns)
filas

In [ ]:
# Guardar en un archivo CSV
postulantes_unicos.to_csv('../data/postulantes_unicos.csv', index=False)  # El argumento 'index=False' evita guardar el índice del DataFrame

## Conclusiones de la Exploracion de Datos

# <a id="1"><font size="6"><div style="border-radius:5px;color:white;background-color:#FA8072;"> 4. EDA Analisis y Exploracion del conjunto de datos 📊 </div> </font>
</a>
<div align="center" style="border: 2px solid #E57373; border-radius: 10px; padding: 20px; background-color: #E3F2FD;">
    <a href="[about:blank](https://www.analyticsvidhya.com/blog/2018/04/introduction-to-graph-theory-network-analysis-python-codes/)" target="_blank">
        <img src="../images/eda.png" alt="Red de Nodos" style="border-radius: 5px; max-width: 80%; box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);">
    </a>
    <p style="font-size: 16px; color: #0a0a0a; margin-top: 10px;">
        <em> <strong>EDA</strong>:
        ¡Imágenes, gráficos y más imágenes!</em>
    </p>
</div>

## A) VARIABLE ESCUELA


### **Descripción**


In [ ]:
postulantes_unicos["ESCUELA"].describe()

<div style="background-color: #f5f5f5; border: 1px solid #ddd; padding: 15px; border-radius: 10px;color: black">
    <p style="font-size: 18px; color: #2E86C1;">Observaciones sobre la Variable "ESCUELA"</p>
    <p>Al analizar la columna <strong>ESCUELA</strong> en el conjunto de datos <strong>postulantes_unicos</strong>, se obtienen las siguientes observaciones:</p>
    <ul>
        <li>En total, existen <strong>3748</strong> registros en la columna, lo que sugiere que hay 3748 postulantes únicos en el conjunto de datos.</li>
        <li>La columna presenta <strong>6</strong> valores únicos, indicando que hay 6 escuelas profesionales distintas a las que los postulantes han aplicado.</li>
        <li>La escuela más frecuente entre los postulantes es <strong>INGENIERÍA CIVIL</strong>, con un total de <strong>1305</strong> apariciones.</li>
        <li>La frecuencia más alta corresponde a la escuela <strong>INGENIERÍA CIVIL</strong>, que aparece en <strong>1305</strong> registros.</li>
    </ul>
    <p>Estas observaciones resaltan la preferencia de los postulantes por la escuela de <strong>INGENIERÍA CIVIL</strong> y proporcionan una visión de la distribución de las preferencias de escuelas entre los postulantes únicos.</p>
</div>

### **GRAFICO DE BARRAS**


In [ ]:
# Contar la frecuencia de cada valor en la columna 'ESCUELA'
escuela_counts = postulantes_unicos["ESCUELA"].value_counts()

# Calcular los porccentajes
total_escuelas = len(
    postulantes_unicos["ESCUELA"]
)  # Usamos Postulantes en lugar de data
porcentajes = (escuela_counts / total_escuelas) * 100

# Crear las etiquetas con porcentajes
etiquetas = [
    f"{escuela} ({porcentaje:.2f} %)"
    for escuela, porcentaje in zip(escuela_counts.index, porcentajes)
]

# Crear una paleta de colores usando seaborn
colors = [
    "hsl(" + str(h) + ",50%" + ",50%)" for h in np.linspace(0, 360, len(escuela_counts))
]

# Crear el gráfico de barras
trace = go.Bar(
    x=etiquetas,  # Usamos las etiquetas con porcentajes como nombres de las barras
    y=escuela_counts.values,  # Frecuencia de cada escuela
    marker=dict(color=colors),  # Color de las barras
)

layout = go.Layout(
    title="Distribución de Escuelas",
    xaxis=dict(title="Escuela"),
    yaxis=dict(title="Frecuencia"),
    paper_bgcolor="rgb(243, 243, 243)",
    plot_bgcolor="rgb(243, 243, 243)",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=[trace], layout=layout)
py.iplot(fig)

**OBSERVACIONES:**

- Se puede observar que las carreras mas populares son INGENIERIA CIVIL y INGENIERIA AGRONOMICA TROPICAL
entre los postulantes, mientras que CONTABILIDAD es la escuela con el menor número de postulantes.

- En base a esto podemos decir que para codificar esta variable podemos considerar lo siguiente:


### **Grafico de Dispersion de la Variable Escuela**

In [ ]:
from dash import Dash, html, dcc, callback, Output, Input
import pandas as pd
import plotly.express as px

# Inicializar la aplicación
app = Dash(__name__)

# Excluir la columna 'ESCUELA'
excluded_columns = ["ESCUELA"]
columns_for_dropdown = [
    col for col in postulantes_unicos.columns if col not in excluded_columns
]

# Diseño de la aplicación
app.layout = html.Div(
    style={"textAlign": "center", "padding": "20px"},
    children=[
        html.H1("Gráfico de Dispersión Interactivo", style={"color": "#2E86C1"}),
        html.Hr(),
        dcc.Dropdown(
            id="dropdown-column",
            options=[{"label": col, "value": col} for col in columns_for_dropdown],
            value=columns_for_dropdown[0],  # Escoge la primera columna por defecto
            multi=False,
            style={"width": "50%", "margin": "auto"},
        ),
        dcc.Graph(id="scatter-plot", style={"margin-top": "20px"}),
    ],
)

# Definir la función de devolución de llamada
@callback(
    Output(component_id="scatter-plot", component_property="figure"),
    Input(component_id="dropdown-column", component_property="value"),
)
def update_scatter_plot(selected_column):
    scatter_data = postulantes_unicos[[selected_column, "ESCUELA"]]
    fig = px.scatter(
        scatter_data,
        x=selected_column,
        y="ESCUELA",
        title=f"Gráfico de Dispersión: {selected_column} vs ESCUELA",
    )
    fig.update_layout(
        plot_bgcolor="#F9F9F9",  # Color de fondo del gráfico
        paper_bgcolor="#F9F9F9",  # Color de fondo del área de trazado
        font_color="#333333",  # Color de fuente
    )
    return fig

# Ejecutar la aplicación
app.run_server(debug=True)


- UUID -> Frecuencia de postulantes de IC -> [1305]
- MODALIDAD -> Frecuencia de MOdalidades IC -> [7]
- AÑO_POSTULACION -> Frecuencia de Año -> [5]


**OBSERVACIONES: **

UUID
- Hay mas postulantes que se presentaron para Ingenieria Civil 
- Hay menos postulantes que se presentaron para Economia y Contabilidad
MODALIDAD
- Unicamente los postulantes de la modalidad ORDINARIO se presentaron a CONTABILIDAD y ECONOMIA
(En general se presnetarona a todas)
- La unica modalidad que no se presento a ECOTURISMO es EXTRAORDINARIO-GRADUADOS Y TITULADOS
AÑO DE POSTULACION
- Al parecer se añadio una nueva carrera de ECONOMIA Y CONTABILIDAD en el 2023 porque solo en este 
año hubo postulantes para esta carrera 

UBIGEO, DEPARTAMENTO, PROVINCIA Y DISTRITO
- Hay varios postulantes de diferenets lugares que postularon a diferentes escuelas

EDAD:
- hay una persona con 67 años de edad esto podria ser un dato atipico
SEXO:
- Hay mujeres y varones que psotularon para cada una de las escuelas

PAGO:
- Hay datos atipicos de 670(ingerieria de alimentos y Civil) y de pago 0(menos en econimia y contabilidad) 




### **Diagrama de Cajas de ESCUELA y PAGO**


In [ ]:
import plotly.graph_objs as go
import plotly.offline as py

# Crear una lista de trazas de cajasC para cada escuela
data_traces = []
for escuela in postulantes_unicos["ESCUELA"].unique():
    trace = go.Box(
        y=postulantes_unicos[postulantes_unicos["ESCUELA"] == escuela][
            "PAGO"
        ],  # Valores a graficar (en este caso, la columna PAGO)
        name=escuela,  # Etiqueta de la caja (escuela)
    )
    data_traces.append(trace)

layout = go.Layout(
    title="Diagrama de Cajas por Escuela Profesional",
    xaxis=dict(title="Escuela Profesional"),
    yaxis=dict(title="Pago"),
    paper_bgcolor="rgb(243, 243, 243)",
    plot_bgcolor="rgb(243, 243, 243)",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=data_traces, layout=layout)
py.iplot(fig)

**Observaciones**

- Datos atipicos
- Datos donde el costo es 0 (Este dato atipico no es muy evidente por lo tanto lo conservamos)
- Datos donde el apgo es 670(Este si e sun dato atipico evidente devido que esta muy alejado del valor medio estandar)


**VALORES ATIPICOS PAGO = 670**


In [ ]:
# Crear un nuevo DataFrame con las filas donde PAGO es igual a cero
postulantes_con_pago_atipico = postulantes_unicos[postulantes_unicos["PAGO"] == 670]

# Mostrar las primeras filas del nuevo DataFrame
postulantes_con_pago_atipico.shape[0]

**Distribución de Postulantes con PAGO=670**


In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Configuración de estilo de Seaborn
sns.set(style="whitegrid")

# Crear un gráfico de barras utilizando 'ESCUELA' en el eje x y 'DISTRITO' en el eje y
plt.figure(figsize=(10, 6))
sns.countplot(data=postulantes_con_pago_atipico, x="ESCUELA", hue="MODALIDAD")
plt.xticks(rotation=90)
plt.title("Distribución de Postulantes con PAGO=670 por Escuela y MOdalidad")
plt.xlabel("Escuela Profesional")
plt.ylabel("Cantidad de Postulantes")
plt.legend(title="Modalidad", loc="upper right")
plt.tight_layout()
plt.show()

**OBSERVACION**
- Los postulantes que pagan 670 son de la modalidad ETRAORDINARIO-GRADUADOS Y TITULADOS

- Sin embargo Existen UNICAMENTE 8 datos atipicos con pago igual a 670, por lo tanto estos datos es mejor de eliminarlos del conjunto de datos


### **Eliminacion de outlayers de la variable PAGO**


In [ ]:
# Crear un nuevo DataFrame sin las filas donde PAGO es igual a 670
postulantes_unicos = postulantes_unicos[postulantes_unicos["PAGO"] != 670]

# Mostrar las primeras filas del nuevo DataFrame
postulantes_unicos.shape[0]

**VALORES ATIPICOS PAGO = 0**


In [ ]:
# Crear un nuevo DataFrame con las filas donde PAGO es igual a cero
postulantes_con_pago_cero = postulantes_unicos[postulantes_unicos["PAGO"] == 0]

# Mostrar las primeras filas del nuevo DataFrame
postulantes_con_pago_cero.shape[0]

**OBSERVACION**

- Existen 45 postulanyes con pago 0, veamos porque a estos postulantes se les excluyo el pago


**Distribución de Postulantes con PAGO=0**


In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Configuración de estilo de Seaborn
sns.set(style="whitegrid")

# Crear un gráfico de barras utilizando 'ESCUELA' en el eje x y 'DISTRITO' en el eje y
plt.figure(figsize=(10, 6))
sns.countplot(data=postulantes_con_pago_cero, x="ESCUELA", hue="MODALIDAD")
plt.xticks(rotation=90)
plt.title("Distribución de Postulantes con PAGO=0 por Escuela y Distrito")
plt.xlabel("Escuela Profesional")
plt.ylabel("Cantidad de Postulantes")
plt.legend(title="Distrito", loc="upper right")
plt.tight_layout()
plt.show()

OBSERVACION

- Los postulantes con modalidad EXTRAORDINARIO VICTIMAS DE TERRORISMO no pagan para rendir el examen


### **Codificacion de la variable ESCUELA - Metodo (Frecuencia()% encoder)**

In [ ]:
# Crear un diccionario a partir de la tupla para mapear las escuelas a los porcentajes
escuela_porcentaje_dict = dict(porcentajes)

# Codificar la variable ESCUELA con los valores enteros de los porcentajes de frecuencia
postulantes_unicos["ESCUELA_CODIFICADA"] = (
    postulantes_unicos["ESCUELA"].map(escuela_porcentaje_dict).apply(lambda x: int(x))
)
postulantes_unicos["ESCUELA_CODIFICADA"]

## B) VARIABLE MODALIDAD


In [ ]:
postulantes_unicos["MODALIDAD"].describe()

In [ ]:
# Contar la frecuencia de cada valor en la columna 'ESCUELA'
escuela_counts = postulantes_unicos["MODALIDAD"].value_counts()

# Calcular los porccentajes
total_escuelas = len(
    postulantes_unicos["MODALIDAD"]
)  # Usamos Postulantes en lugar de data
porcentajes = (escuela_counts / total_escuelas) * 100

# Crear las etiquetas con porcentajes
etiquetas = [
    f"{escuela} ({porcentaje:.2f} %)"
    for escuela, porcentaje in zip(escuela_counts.index, porcentajes)
]

# Crear una paleta de colores usando seaborn
colors = [
    "hsl(" + str(h) + ",50%" + ",50%)" for h in np.linspace(0, 360, len(escuela_counts))
]

# Crear el gráfico de barras
trace = go.Bar(
    x=etiquetas,  # Usamos las etiquetas con porcentajes como nombres de las barras
    y=escuela_counts.values,  # Frecuencia de cada escuela
    marker=dict(color=colors),  # Color de las barras
)

layout = go.Layout(
    title="Distribución de Modalidades",
    xaxis=dict(title="Modalidad"),
    yaxis=dict(title="Frecuencia"),
    paper_bgcolor="rgb(243, 243, 243)",
    plot_bgcolor="rgb(243, 243, 243)",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=[trace], layout=layout)
py.iplot(fig)

### **Codificacion de la variable MODALIDAD**

In [ ]:
# Crear un diccionario a partir de la tupla para mapear las escuelas a los porcentajes
escuela_porcentaje_dict = dict(porcentajes)

# Codificar la variable ESCUELA con los valores enteros de los porcentajes de frecuencia
postulantes_unicos["MODALIDAD_CODIFICADA"] = (
    postulantes_unicos["MODALIDAD"].map(escuela_porcentaje_dict).apply(lambda x: int(x))
)
postulantes_unicos["MODALIDAD_CODIFICADA"]

###  **Diagrama de Cajas Modalidada**


In [ ]:
import plotly.graph_objs as go
import plotly.offline as py

# Crear una lista de trazas de cajasC para cada escuela
data_traces = []
for escuela in postulantes_unicos["MODALIDAD"].unique():
    trace = go.Box(
        y=postulantes_unicos[postulantes_unicos["MODALIDAD"] == escuela][
            "PAGO"
        ],  # Valores a graficar (en este caso, la columna PAGO)
        name=escuela,  # Etiqueta de la caja (escuela)
    )
    data_traces.append(trace)

layout = go.Layout(
    title="Diagrama de Cajas por Modalidad",
    xaxis=dict(title="Modalidad"),
    yaxis=dict(title="Pago"),
    paper_bgcolor="rgb(243, 243, 243)",
    plot_bgcolor="rgb(243, 243, 243)",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=data_traces, layout=layout)
py.iplot(fig)

In [ ]:
import plotly.graph_objs as go
import plotly.offline as py

# Filtrar por sexo (por ejemplo, 'MASCULINO' o 'FEMENINO')
sexo_filtrado = "MASCULINO"

# Filtrar el DataFrame por el sexo seleccionado
data_sexo_filtrado = postulantes_unicos[postulantes_unicos["SEXO"] == sexo_filtrado]

# Crear una lista de trazas de dispersión para cada escuela
data_traces = []
for escuela in data_sexo_filtrado["MODALIDAD"].unique():
    trace = go.Scatter(
        x=data_sexo_filtrado[data_sexo_filtrado["MODALIDAD"] == escuela]["MODALIDAD"],
        y=data_sexo_filtrado[data_sexo_filtrado["MODALIDAD"] == escuela]["EDAD"],
        mode="markers",
        name=escuela,
    )
    data_traces.append(trace)

layout = go.Layout(
    title="Distribución de Edad por Escuela Profesional (Filtro: "
    + sexo_filtrado
    + ")",
    xaxis=dict(title="Escuela Profesional"),
    yaxis=dict(title="Edad"),
    paper_bgcolor="rgb(243, 243, 243)",
    plot_bgcolor="rgb(243, 243, 243)",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=data_traces, layout=layout)
py.iplot(fig)

### **Eliminacion de los valores atipicos con Edad 67**

In [ ]:
# Filtrar el conjunto de datos para obtener las filas con EDAD igual a 67
postulantes_edad_67 = postulantes_unicos[postulantes_unicos["EDAD"] == 67]

# Filtrar las filas restantes con EDAD diferente de 67 para eliminar los valores atípicos
postulantes_unicos = postulantes_unicos[postulantes_unicos["EDAD"] != 67]

# Mostrar el tamaño del conjunto de datos sin valores atípicos
print("Tamaño del conjunto de datos sin valores atípicos:", postulantes_unicos.shape)


## C) VARIABLE IDIOMA


In [ ]:
import plotly.graph_objs as go
import plotly.offline as py

# Filtrar las columnas de interés
data_filtered = postulantes_unicos[["ESCUELA", "IDIOMA"]]

# Contar la cantidad de estudiantes por combinación de Escuela e Idioma
idioma_counts = (
    data_filtered.groupby(["ESCUELA", "IDIOMA"]).size().reset_index(name="Cantidad")
)

# Crear el gráfico de barras
data = []
for idioma in idioma_counts["IDIOMA"].unique():
    trace = go.Bar(
        x=idioma_counts[idioma_counts["IDIOMA"] == idioma]["ESCUELA"],
        y=idioma_counts[idioma_counts["IDIOMA"] == idioma]["Cantidad"],
        name=idioma,
    )
    data.append(trace)

layout = go.Layout(
    title="Distribución de Idiomas por Escuela Profesional",
    xaxis=dict(title="Escuela Profesional"),
    yaxis=dict(title="Cantidad"),
    barmode="stack",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=data, layout=layout)
py.iplot(fig)

In [ ]:
import plotly.graph_objs as go
import plotly.offline as py

# Contar la frecuencia de cada valor en la columna 'IDIOMA'
idioma_counts = postulantes_unicos["IDIOMA"].value_counts()

# Crear el gráfico de pastel en 3D
trace = go.Pie(
    labels=idioma_counts.index,
    values=idioma_counts.values,
    type="pie",
    textinfo="label+percent",
    hoverinfo="label+value+percent",
    textfont=dict(size=12),
    marker=dict(
        colors=["#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b"]
    ),
)

layout = go.Layout(
    title="Distribución de Idiomas",
    paper_bgcolor="rgb(243, 243, 243)",
    plot_bgcolor="rgb(243, 243, 243)",
)

# Crear la figura y trazar el gráfico
fig = go.Figure(data=[trace], layout=layout)
py.iplot(fig)

In [ ]:
idioma_counts

### **Codificacion de la variable IDIOMA con el metodo - Frecuencia(%) encoding**

In [ ]:
import pandas as pd

# Calcular los porcentajes
idioma_counts = postulantes_unicos["IDIOMA"].value_counts()
total_registros = len(postulantes_unicos["IDIOMA"])
porcentajes = (idioma_counts / total_registros) * 100

# Crear un diccionario a partir de la serie para mapear los idiomas a los porcentajes
idioma_porcentaje_dict = porcentajes.to_dict()


# Codificar la variable IDIOMA con los valores enteros de los porcentajes de frecuencia
def codificar_idioma(idioma):
    if idioma in idioma_porcentaje_dict:
        return int(idioma_porcentaje_dict[idioma])
    else:
        return None


postulantes_unicos["IDIOMA_CODIFICADO"] = postulantes_unicos["IDIOMA"].apply(
    codificar_idioma
)

postulantes_unicos[["IDIOMA", "IDIOMA_CODIFICADO"]]

In [ ]:
import pandas as pd

# Filtrar los datos para aquellos con información de idioma
data_with_idioma = postulantes_unicos[postulantes_unicos["IDIOMA"].notnull()]

# Agrupar por carrera y contar cuántos hablan cada idioma
idioma_counts_by_carrera = (
    data_with_idioma.groupby(["ESCUELA", "IDIOMA"])["UUID"].count().reset_index()
)

# Renombrar la columna UUID por 'Cantidad'
idioma_counts_by_carrera.rename(columns={"UUID": "Cantidad"}, inplace=True)

# Mostrar la tabla
idioma_counts_by_carrera.head()

In [ ]:
import pandas as pd

# Filtrar los datos para aquellos con información de idioma
data_with_idioma = postulantes_unicos[postulantes_unicos["IDIOMA"].notnull()]

# Agrupar por carrera y contar cuántos hablan cada idioma
idioma_counts_by_carrera = (
    data_with_idioma.groupby(["ESCUELA", "IDIOMA"])["UUID"].count().reset_index()
)

# Renombrar la columna UUID por 'Cantidad'
idioma_counts_by_carrera.rename(columns={"UUID": "Cantidad"}, inplace=True)

# Ordenar la tabla por cantidad y por escuela profesional
idioma_counts_by_carrera_sorted = idioma_counts_by_carrera.sort_values(
    by=["Cantidad", "ESCUELA"], ascending=[False, True]
)

# Mostrar la tabla ordenada
idioma_counts_by_carrera_sorted

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Filtrar las columnas de interés
data_filtered = postulantes_unicos[["ESCUELA", "IDIOMA"]]

# Crear el gráfico de barras
plt.figure(figsize=(12, 6))
sns.countplot(x="ESCUELA", hue="IDIOMA", data=data_filtered)
plt.xticks(rotation=90)
plt.title("Distribución de Idiomas por Escuela Profesional")
plt.xlabel("Escuela Profesional")
plt.ylabel("Cantidad")
plt.legend(title="Idioma")
plt.tight_layout()
plt.show()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Filtrar los datos para aquellos con información de idioma y UBIGEO_IE
data_with_idioma_ubigeo = postulantes_unicos[
    postulantes_unicos["IDIOMA"].notnull() & (postulantes_unicos["UBIGEO_IE"] != 0)
]

# Crear el gráfico de dispersión utilizando Seaborn
plt.figure(figsize=(12, 6))
sns.scatterplot(data=data_with_idioma_ubigeo, x="UBIGEO_IE", y="IDIOMA", alpha=0.5)
plt.title("Distribución de Idioma por UBIGEO_IE")
plt.xlabel("UBIGEO_IE")
plt.ylabel("Idioma")
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Filtrar los datos para aquellos con información de idioma y PROVINCIA_IE
data_with_idioma_provincia = postulantes_unicos[
    postulantes_unicos["IDIOMA"].notnull() & (postulantes_unicos["PROVINCIA_IE"] != "")
]

# Crear el gráfico de dispersión utilizando Seaborn
plt.figure(figsize=(12, 6))
sns.scatterplot(
    data=data_with_idioma_provincia,
    x="PROVINCIA_IE",
    y="IDIOMA",
    hue="IDIOMA",
    palette="Set1",
    alpha=0.7,
)
plt.title("Distribución de Idioma por PROVINCIA_IE")
plt.xlabel("PROVINCIA_IE")
plt.ylabel("Idioma")
plt.xticks(rotation=90)
plt.tight_layout()
plt.legend(title="Idioma")
plt.show()

In [ ]:
import pandas as pd

# Filtrar los datos para aquellos con información de idioma y PROVINCIA_IE
data_with_idioma_provincia = postulantes_unicos[
    postulantes_unicos["IDIOMA"].notnull() & (postulantes_unicos["PROVINCIA_IE"] != "")
]

# Agrupar los datos por región y idioma, y contar la cantidad de postulantes en cada grupo
idioma_por_region = (
    data_with_idioma_provincia.groupby(["PROVINCIA_IE", "IDIOMA"])["UUID"]
    .count()
    .reset_index()
)
idioma_por_region.columns = ["Region", "Idioma", "Cantidad"]

# Ordenar la tabla por cantidad y región
idioma_por_region = idioma_por_region.sort_values(
    by=["Region", "Cantidad"], ascending=[True, False]
)

# Imprimir la tabla
print(idioma_por_region)

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Crear un gráfico de barras agrupadas para contar la cantidad de postulantes por idioma y provincia
plt.figure(figsize=(12, 8))
sns.set_palette("pastel")
sns.set(style="whitegrid")
sns.countplot(data=data_with_idioma_provincia, x="PROVINCIA_IE", hue="IDIOMA")
plt.title("Cantidad de Postulantes por Idioma y Provincia")
plt.xlabel("Provincia")
plt.ylabel("Cantidad de Postulantes")
plt.xticks(rotation=45)
plt.legend(title="Idioma")
plt.tight_layout()
plt.show()

In [ ]:
import plotly.graph_objs as go
import plotly.offline as py

# Agrupar los datos por provincia e idioma y contar la cantidad de postulantes en cada combinación
grouped_data = (
    data_with_idioma_provincia.groupby(["PROVINCIA_IE", "IDIOMA"])
    .size()
    .reset_index(name="CANTIDAD")
)

# Crear un gráfico de barras agrupadas
trace = []
idiomas = grouped_data["IDIOMA"].unique()
for idioma in idiomas:
    trace.append(
        go.Bar(
            x=grouped_data[grouped_data["IDIOMA"] == idioma]["PROVINCIA_IE"],
            y=grouped_data[grouped_data["IDIOMA"] == idioma]["CANTIDAD"],
            name=idioma,
        )
    )

layout = go.Layout(
    title="Cantidad de Postulantes por Idioma y Provincia",
    xaxis=dict(title="Provincia"),
    yaxis=dict(title="Cantidad de Postulantes"),
    barmode="group",
)

fig = go.Figure(data=trace, layout=layout)
py.iplot(fig)

In [ ]:
correlation_matrix = postulantes_unicos.corr()
sns.heatmap(correlation_matrix, annot=True, cmap="coolwarm")
plt.show()

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Crear la tabla de contingencia entre IDIOMA y DISTRITO_IE
contingency_table = pd.crosstab(
    postulantes_unicos["IDIOMA"], postulantes_unicos["DISTRITO_IE"]
)

# Generar el heatmap
plt.figure(figsize=(10, 6))
sns.heatmap(contingency_table, annot=True, cmap="YlGnBu", fmt="d")
plt.title("Tabla de Contingencia y Heatmap entre IDIOMA y DISTRITO_IE")
plt.xlabel("DISTRITO_IE")
plt.ylabel("IDIOMA")
plt.show()

# CODIFICACION DE LAS DEMAS VARIABLES CATEGORICAS


In [ ]:
!pip install category_encoders

In [ ]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from category_encoders import CountEncoder

# Lista de columnas con variables categóricas y de interés
categorical_cols = [
    "DEPARTAMENTO",
    "PROVINCIA",
    "DISTRITO",
    "DEPARTAMENTO_ORIGEN",
    "PROVINCIA_ORIGEN",
    "DISTRITO_ORIGEN",
    "DEPARTAMENTO_IE",
    "PROVINCIA_IE",
    "DISTRITO_IE",
    "SEXO",
    "LEE",
    "ESCRIBE",
    "HABLA",
    "TIENE_DISCAPACIDAD",
    "TIPO_COMUNIDAD",
    "COMUNIDAD",
]

# Crear un nuevo DataFrame para las variables codificadas
encoded_data = postulantes_unicos.copy()

# Inicializar el LabelEncoder
label_encoder = LabelEncoder()

# Inicializar el CountEncoder
count_encoder = CountEncoder()

# Aplicar Label-Encoding a las variables categóricas con 2 a 5 valores únicos
# y Count-Encoding a las demás
for col in categorical_cols:
    unique_values = postulantes_unicos[col].nunique()
    if unique_values == 2:
        encoded_data[col] = label_encoder.fit_transform(postulantes_unicos[col])
    else:
        encoded_data[col] = count_encoder.fit_transform(postulantes_unicos[col])

## DATAFRAME CODIFICADO

In [ ]:
# Mostrar las primeras filas del DataFrame codificado
encoded_data.head()

## COLUMNAS DEL DATASET CODIFICADO


In [ ]:
encoded_data.columns

## MATRIZ DE CORRELACION

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Calcular la matriz de correlación
correlation_matrix = encoded_data.corr()

# Crear un heatmap de la matriz de correlación
plt.figure(figsize=(12, 10))
sns.heatmap(correlation_matrix, annot=True, cmap="coolwarm", fmt=".1f")
plt.title("Matriz de Correlación")
plt.show()

# Correlación de la variable IDIOMA con otras variables
correlation_with_idioma = correlation_matrix["IDIOMA_CODIFICADO"].sort_values(
    ascending=False
)
print(correlation_with_idioma)

La variable 'IDIOMA_CODIFICADO' tiene una correlación positiva más fuerte con las siguientes variables:

PAGO: 0.651244
TIPO_COMUNIDAD: 0.494760
MODALIDAD_CODIFICADA: 0.431830
DISTRITO_ORIGEN: 0.351687
DISTRITO_IE: 0.332492
DISTRITO: 0.296624
LEE: 0.078223
ESCRIBE: 0.067700


In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

# Calcular la correlación de la variable IDIOMA con las demás variables
correlation_with_idioma = encoded_data.corr()["IDIOMA_CODIFICADO"].drop(
    "IDIOMA_CODIFICADO"
)

# Crear un gráfico de barras de la correlación
plt.figure(figsize=(10, 8))
sns.barplot(
    x=correlation_with_idioma.values,
    y=correlation_with_idioma.index,
    palette="coolwarm",
)
plt.xlabel("Correlación con IDIOMA")
plt.ylabel("Variables")
plt.title("Correlación de IDIOMA con Otras Variables")
plt.show()

In [ ]:
# Verificar los valores faltantes en cada columna
missing_values = postulantes_unicos.isnull().sum()

# Mostrar el total de valores faltantes por columna
print("Valores faltantes por columna:")
print(missing_values)

# IMPUTACION DE LOS DATOS

In [ ]:
# Calcular la moda de las variables seleccionadas
moda_lee = encoded_data["LEE"].mode()[0]
moda_escribe = encoded_data["ESCRIBE"].mode()[0]
moda_habla = encoded_data["HABLA"].mode()[0]
moda_tiene_discapacidad = encoded_data["TIENE_DISCAPACIDAD"].mode()[0]
moda_tipo_comunidad = encoded_data["TIPO_COMUNIDAD"].mode()[0]

# Imputar los valores faltantes con la moda
encoded_data["LEE"].fillna(moda_lee, inplace=True)
encoded_data["ESCRIBE"].fillna(moda_escribe, inplace=True)
encoded_data["HABLA"].fillna(moda_habla, inplace=True)
encoded_data["TIENE_DISCAPACIDAD"].fillna(moda_tiene_discapacidad, inplace=True)
encoded_data["TIPO_COMUNIDAD"].fillna(moda_tipo_comunidad, inplace=True)

# Mostrar las primeras filas del DataFrame con los valores imputados
print(
    encoded_data[
        ["LEE", "ESCRIBE", "HABLA", "TIENE_DISCAPACIDAD", "TIPO_COMUNIDAD"]
    ].head()
)

In [ ]:
import numpy as np
from sklearn.impute import KNNImputer

# Crear una matriz con las variables que se utilizarán para imputar
variables_imputacion = encoded_data[
    [
        "PAGO",
        "TIPO_COMUNIDAD",
        "MODALIDAD_CODIFICADA",
        "DISTRITO_ORIGEN",
        "DISTRITO_IE",
        "DISTRITO",
        "LEE",
        "ESCRIBE",
    ]
].values

# Crear una instancia de KNNImputer
imputer = KNNImputer(
    n_neighbors=5
)  # Puedes ajustar el número de vecinos (neighbors) según tus necesidades

# Imputar los valores faltantes en 'IDIOMA_CODIFICADO'
encoded_data["IDIOMA_CODIFICADO_IMPUTADO"] = imputer.fit_transform(
    variables_imputacion
)[:, 0]

# Mostrar las primeras filas del DataFrame con los valores imputados
print(encoded_data[["IDIOMA_CODIFICADO", "IDIOMA_CODIFICADO_IMPUTADO"]].head())

In [ ]:
encoded_data["IDIOMA_CODIFICADO_IMPUTADO"]

In [ ]:
encoded_data.head()

In [ ]:
columnas = encoded_data.columns
print(columnas)

In [ ]:
columnas = ["ESCUELA", "MODALIDAD", "IDIOMA", "IDIOMA_CODIFICADO"]
DfEstudiantes = encoded_data.drop(columns=columnas)

# DATASET PREPROCESADO

In [ ]:
DfEstudiantes.head()

In [ ]:
DfEstudiantes.columns

In [ ]:
DfEstudiantes.describe()

In [ ]:
# Eliminar la columna 'Fecha_Corte'
DfEstudiantes = DfEstudiantes.drop(columns=['UBIGEO','UBIGEO_ORIGEN','UBIGEO_IE'])

# Mostrar el conjunto de datos después de eliminar la columna
DfEstudiantes.head()

In [ ]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler

# Crear una copia del DataFrame para evitar modificar el original
df_normalized = DfEstudiantes.copy()
df_standardized = DfEstudiantes.copy()

# Seleccionar las columnas numéricas para normalizar y estandarizar
numeric_columns = [
    "ANNIO_POSTULACION",
    "DEPARTAMENTO",
    "PROVINCIA",
    "DISTRITO",
    "DEPARTAMENTO_ORIGEN",
    "PROVINCIA_ORIGEN",
    "DISTRITO_ORIGEN",
    "PAGO",
    "LEE",
    "ESCRIBE",
    "HABLA",
    "TIENE_DISCAPACIDAD",
    "TIPO_COMUNIDAD",
    "COMUNIDAD",
    "ESCUELA_CODIFICADA",
    "MODALIDAD_CODIFICADA",
    "IDIOMA_CODIFICADO_IMPUTADO",
]

# Normalizar los datos
scaler = MinMaxScaler()
df_normalized[numeric_columns] = scaler.fit_transform(df_normalized[numeric_columns])

# Estandarizar los datos
scaler = StandardScaler()
df_standardized[numeric_columns] = scaler.fit_transform(
    df_standardized[numeric_columns]
)

# Mostrar las primeras filas de los DataFrames normalizados y estandarizados
print("DataFrame Normalizado:")
print(df_normalized.head())

print("\nDataFrame Estandarizado:")
print(df_standardized.head())

In [ ]:
DfEstudiantes.describe()

In [ ]:
#DfEstudiantes['DEPARTAMENTO'].unique()
#for i in DfEstudiantes:
#    c = 1
#    arr = DfEstudiantes[i].unique()
#    newdf = DfEstudiantes[i][j]
#   for j in arr:
#        newdf[i][j] = c
#       c += 1
        
# Supongamos que tienes un DataFrame DfEstudiantes con columnas

# Crear un DataFrame vacío para almacenar los valores numerizados
numerized_df = pd.DataFrame()

# Iterar a través de las columnas del DataFrame original
for column in DfEstudiantes.columns:
    unique_values = DfEstudiantes[column].unique()
    mapping = {value: index + 1 for index, value in enumerate(unique_values)}
    numerized_column = DfEstudiantes[column].map(mapping)
    numerized_df[column] = numerized_column

print(numerized_df)

In [ ]:
import numpy as np

# Convertir el DataFrame a un array numpy
array_estudiantes = numerized_df.values

print(array_estudiantes)

### Creamos la matriz de correlacion de estudiantes


#### MaTRIZ DE CORRELACION UTILIZANDO LA DISTACIA COSENO


In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Supongamos que DfEstudiantes contiene solo las columnas numéricas (sin UUID)
array_estudiantes = DfEstudiantes.drop(columns=["UUID"]).values

# Calcular la matriz de similitud (correlación) entre los postulantes utilizando la distancia coseno
similarity_matrix = np.dot(array_estudiantes, array_estudiantes.T) / np.outer(
    np.linalg.norm(array_estudiantes, axis=1), np.linalg.norm(array_estudiantes, axis=1)
)

# Crear un DataFrame para la matriz de similitud
column_names = DfEstudiantes["UUID"].tolist()
similarity_df = pd.DataFrame(
    similarity_matrix, columns=column_names, index=column_names
)

In [ ]:
similarity_df

In [ ]:
# Obtener las primeras 5 filas y 5 columnas del conjunto de datos similarity_df
primeros_5 = similarity_df.iloc[30:50, 30:50]

# Mostrar la matriz de correlación en forma de mapa de calor
plt.figure(figsize=(16, 12))
sns.heatmap(primeros_5, annot=True, cmap="coolwarm", fmt=".5f")
plt.title("Matriz de Correlación de Similitud entre los Primeros 5 Postulantes")
plt.show()

#### MATRIZ DE CORRELACION UTILIZANDO LA CORRELACION DE PEARSON


In [ ]:
import pandas as pd
import numpy as np

# Supongamos que DfEstudiantes contiene solo las columnas numéricas (sin UUID)
numeric_columns = DfEstudiantes.drop(columns=["UUID"])

# Calcular la matriz de correlación de Pearson entre las filas (postulantes)
correlation_matrix = numeric_columns.T.corr(method="pearson")

# Obtener solo la parte triangular superior de la matriz de correlación
upper_triangle = np.triu(correlation_matrix, k=1)

# Visualizar la matriz de correlación de Pearson
upper_triangle

In [ ]:
similarity_df = pd.DataFrame(
    upper_triangle, columns=column_names, index=column_names
)
similarity_df

In [ ]:
similarity_df.iloc[1, 3]

##  Muestra de la matriz de correlacion

In [ ]:
import pandas as pd

# Crear una muestra aleatoria de 5 filas y 5 columnas del DataFrame de correlación
sampled_correlation_df = similarity_df.sample(n=1000, axis=0).sample(n=1000, axis=1)

# Mostrar la muestra aleatoria
print(sampled_correlation_df)

In [ ]:
import networkx as nx

# Crear una instancia del grafo
G = nx.Graph()
similarity_df = sampled_correlation_df
# Obtener los nombres de los postulantes (nodos)
postulantes = similarity_df.columns.tolist()

n=1000
# Agregar nodos al grafo
G.add_nodes_from(postulantes[:])

th = 0.80

# Crear un conjunto para rastrear las aristas ya agregadas
added_edges = set()

# Agregar aristas al grafo con los valores de similitud sin repeticiones
for i in range(len(postulantes[:])):
    for j in range(i + 1, len(postulantes[:])):
        # Filtro con el Threshold
        if abs(similarity_df.iloc[i, j]) >= th:
            # Ordenar nodos para asegurarse de que estén en el mismo orden
            nodes = tuple(sorted([postulantes[i], postulantes[j]]))
            if nodes not in added_edges:
                dist = 1 - similarity_df.iloc[i, j]
                G.add_edge(*nodes, weight=dist)
                added_edges.add(nodes)

# Mostrar información básica sobre el grafo
print("Número de nodos:", G.number_of_nodes())
print("Número de aristas:", G.number_of_edges())


In [ ]:
# Eliminar aristas que son bucles
self_loop_edges = list(nx.selfloop_edges(G))
G.remove_edges_from(self_loop_edges)

# Mostrar información básica sobre el grafo
print("Número de nodos:", G.number_of_nodes())
print("Número de aristas:", G.number_of_edges())


In [ ]:
import random

# Convertir el NodeView a una lista
node_list = list(G.nodes())

# Realizar el muestreo del grafo
sample_size = 300  # Número de nodos en la muestra
sample = random.sample(node_list, sample_size)

# Crear una submuestra del grafo original
sampled_graph = G.subgraph(sample)

# Mostrar información básica sobre la submuestra
print("Número de nodos en la submuestra:", sampled_graph.number_of_nodes())
print("Número de aristas en la submuestra:", sampled_graph.number_of_edges())


In [ ]:
import networkx as nx
import matplotlib.pyplot as plt

# Crear una disposición de nodos para la visualización
pos = nx.spring_layout(G, seed=42)

# Obtener los pesos de las aristas para ajustar el ancho de las aristas en la visualización
edge_weights = [data['weight'] for _, _, data in G.edges(data=True)]

# Crear el grafo utilizando la disposición de nodos y los pesos de las aristas
nx.draw(G, pos, with_labels=True, node_size=100, node_color='skyblue', font_size=2, font_color='black',
        width=edge_weights, edge_color='gray', font_weight='bold')
# Ajustar el tamaño del gráfico
# Mostrar el grafo
plt.title("Grafo de Muestra de Similitud entre Postulantes")
plt.show()


In [ ]:
import networkx as nx
import matplotlib.pyplot as plt

# ... (código para construir el grafo G)

plt.figure(figsize=(30, 30))
spring_layout = nx.spring_layout(sampled_graph, seed=42)

nx.draw_networkx_nodes(sampled_graph, spring_layout, node_color='r', edgecolors='k')
nx.draw_networkx_edges(sampled_graph, spring_layout, alpha=0.7)

plt.title('Graph with spring layout', fontdict={'fontsize': 40})
plt.axis('off')
plt.show()

In [ ]:
# Parámetros de visualización
plt.figure(figsize=(10,10))
pos = nx.spring_layout(sampled_graph)
nx.draw_networkx_nodes(sampled_graph, pos, node_size=50)
nx.draw_networkx_edges(sampled_graph, pos)
nx.draw_networkx_labels(sampled_graph, pos)

# Título y ejes
plt.title('Red de Similitudes entre Postulantes')
plt.axis('off')
plt.show()

In [ ]:
import networkx as nx
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.decomposition import PCA
from node2vec import Node2Vec as n2v
sns.set()
WINDOW = 1  # Node2Vec fit window
MIN_COUNT = 1  # Node2Vec min. count
BATCH_WORDS = 4  # Node2Vec batch words

g_emb = n2v(
    G,
    dimensions=16
)

mdl = g_emb.fit(
    vector_size=16,
    window=WINDOW,
    min_count=MIN_COUNT,
    batch_words=BATCH_WORDS
)

In [ ]:
emb_df = (
    pd.DataFrame(
        [mdl.wv.get_vector(str(n)) for n in G.nodes()],
        index=G.nodes
    )
)

In [ ]:

pca = PCA(n_components=2, random_state=7)
pca_mdl = pca.fit_transform(emb_df)

emb_df_PCA = (
    pd.DataFrame(
        pca_mdl,
        columns=['x', 'y'],
        index=emb_df.index
    )
)
plt.clf()
fig = plt.figure(figsize=(6, 4))
plt.scatter(
    x=emb_df_PCA['x'],
    y=emb_df_PCA['y'],
    s=0.4,
    color='maroon',
    alpha=0.5
)
plt.xlabel('PCA-1')
plt.ylabel('PCA-2')
plt.title('PCA Visualization')
plt.plot()
